projects/packages/autoloader/tests/php/lib/class-test-plugin-factory.php

```diff
@@ -0,0 +1,530 @@
+	 * @return Test_Plugin_Factory
+	 * @throws \InvalidArgumentException When the slug is invalid.
+	 */
+	public static function create( $is_mu_plugin, $slug, $autoloads ) {
+		if ( false !== strpos( $slug, ' ' ) ) {
+			throw new \InvalidArgumentException( 'Plugin slugs may not have spaces.' );
+		}
+
+		$slug = strtolower( preg_replace( '/[^A-Za-z0-9\-_]/', '', $slug ) );
+
+		return new Test_Plugin_Factory( $is_mu_plugin, $slug, $autoloads );
+	}
+
+	/**
+	 * Creates a new factory configured for a generic test plugin and returns it.
+	 *
+	 * @param bool   $is_mu_plugin Indicates whether or not the plugin is an mu-plugin.
+	 * @param string $version      The version of the autoloader we want the plugin to use.
+	 * @return Test_Plugin_Factory
+	 */
+	public static function create_test_plugin( $is_mu_plugin, $version ) {
+		// We will use a global to detect when a file has been loaded by the autoloader.
+		global $jetpack_autoloader_testing_loaded_files;
+		if ( ! isset( $jetpack_autoloader_testing_loaded_files ) ) {
+			$jetpack_autoloader_testing_loaded_files = array();
+		}
+
+		$file_version = $version;
+		if ( self::CURRENT === $version ) {
+			$file_version      = self::VERSION_CURRENT;
+			$namespace_version = 'Current';
+		} else {
+			$namespace_version = 'v' . str_replace( '.', '_', $version );
+		}
+
+		// Avoid namespace collisions between plugins & mu-plugins.
+		if ( $is_mu_plugin ) {
+			$namespace_version .= 'mu';
+		}
+
+		// We need to define all of the autoloads that the files contained within will utilize.
+		$autoloads = array(
+			'classmap' => array( 'includes' ),
+			'psr-4'    => array(
+				self::TESTING_NAMESPACE => 'src',
+			),
+			'files'    => array( 'functions.php' ),
+		);
+
+		return self::create( $is_mu_plugin, str_replace( '.', '_', $version ), $autoloads )
+			->with_class( 'classmap', 'Classmap_Test_Class', "\tconst VERSION = '$file_version';" )
+			->with_class( 'psr-4', self::TESTING_NAMESPACE . 'SharedTestClass', "\tconst VERSION = '$file_version';" )
+			->with_class( 'psr-4', self::TESTING_NAMESPACE . "$namespace_version\\UniqueTestClass", '' )
+			->with_file( 'functions.php', "<?php\n\nglobal \$jetpack_autoloader_testing_loaded_files;\n\$jetpack_autoloader_testing_loaded_files[] = '$file_version';" )
+			->with_autoloader_version( $version )
+			->with_composer_config( array( 'config' => array( 'autoloader-suffix' => $namespace_version ) ) );
+	}
+
+	/**
+	 * Adds a file to the plugin being built.
+	 *
+	 * @param string $path    The path for the file in the plugin directory.
+	 * @param string $content The content for the file.
+	 * @return $this
+	 */
+	public function with_file( $path, $content ) {
+		$this->files[ $path ] = $content;
+		return $this;
+	}
+
+	/**
+	 * Adds a class file to the plugin being built.
+	 *
+	 * @param string $autoload_type The type of autoloading to name the fule for. 'classmap', 'psr-4', 'psr-0' are options.
```

Comment:
I think this is a typo:
```suggestion
	 * @param string $autoload_type The type of autoloading to name the file for. 'classmap', 'psr-4', 'psr-0' are options.
```